### PR TITLE
Url and Address Url set in TrainedModel Status using InferenceService Status

### PR DESCRIPTION
**Pull request by abchoo**
**What this PR does / why we need it**: Included url and addressable url in TrainedModel from parent inference service. Test are added to confirm proper updates.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubeflow/kfserving/issues/1007

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/controller/v1alpha1/trainedmodel/controller.go [hunk 1]
```diff
@@ -118,15 +122,78 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
 
-func (r *TrainedModelReconciler) updateStatus(desiredService *v1alpha1api.TrainedModel) error {
-	//TODO update TrainedModel status object, this will be done in a separate PR
+func (r *TrainedModelReconciler) updateStatus(req ctrl.Request, desiredModel *v1alpha1api.TrainedModel) error {
+	// Get the parent inference service
+	isvc := &v1beta1api.InferenceService{}
+	if err := r.Get(context.TODO(), types.NamespacedName{Namespace: req.Namespace, Name: desiredModel.Spec.InferenceService}, isvc); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("Parent InferenceService does not exists, deleting TrainedModel", "TrainedModel", desiredModel.Name, "InferenceService", isvc.Name)
+			r.Delete(context.TODO(), desiredModel)
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/trainedmodel/controller.go:143]: I think we shouldn't delete trained model here, instead we can return error back and the next reconciliation loop will retry.

File: pkg/controller/v1alpha1/trainedmodel/controller.go [hunk 2]
```diff
@@ -118,15 +122,78 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		return err
+	}
+
+	// obtain protocol version of predictor spec
+	protocolVersion := string(isvc.Spec.Predictor.GetImplementation().GetProtocol())
+
+	// Check if parent inference service has the status URL
+	if isvc.Status.URL != nil {
+		// Update status to contain the isvc URL with /v1/models/trained-model-name:predict appened
+		url := isvc.Status.URL.String() + "/" + protocolVersion + "/models/" + desiredModel.Name + ":predict"
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/trainedmodel/controller.go:155]: for v2 the verb is `/infer`

File: pkg/controller/v1alpha1/trainedmodel/controller.go [hunk 3]
```diff
@@ -118,15 +122,78 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+			return err
+		}
+		desiredModel.Status.URL = externURL
+	}
+
+	// Check if parent inference service has the address URL
+	if isvc.Status.Address != nil {
+		if isvc.Status.Address.URL != nil {
+			////Update status to contain the isvc address with /v1/models/trained-model-name:predict appened
+			url := isvc.Status.Address.URL.String() + "/" + protocolVersion + "/models/" + desiredModel.Name + ":predict"
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/trainedmodel/controller.go:167]: `/infer` for v2 protocol

File: pkg/apis/serving/v1alpha1/trained_model_status.go [hunk 4]
```diff
@@ -17,13 +17,17 @@ limitations under the License.
...
+	"knative.dev/pkg/apis"
 	duckv1 "knative.dev/pkg/apis/duck/v1"
 )
 
 // TrainedModelStatus defines the observed state of TrainedModel
 type TrainedModelStatus struct {
 	// Conditions for trained model
 	duckv1.Status `json:",inline"`
+	// URL holds the url that will distribute traffic over the provided traffic targets.
+	// http[s]://{route-name}.{route-namespace}.{cluster-level-suffix}/v1/models/<trainedmodel>:predict
```
Comment (yuzliu) [inline pkg/apis/serving/v1alpha1/trained_model_status.go:29]: For v2 it should be http[s]://{route-name}.{route-namespace}.{cluster-level-suffix}/v2/models/<trainedmodel>:infer

**Comment by yuzliu** [comment]
I think we can update the e2e test's predict function [here](https://github.com/kubeflow/kfserving/blob/aed797cce7c29779cbe6e056fc1db52b515753a7/test/e2e/common/utils.py#L31) to take trainedmodel as an input and resolve the url then send a request

File: pkg/controller/v1alpha1/trainedmodel/controller.go [hunk 5]
```diff
@@ -118,15 +122,79 @@ func (r *TrainedModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+	// V1 ends with :predict whereas V2 ends with /infer
+	endpointSuffix := ":predict"
+	if protocolVersion == string(constants.ProtocolV2) {
+		endpointSuffix = "/infer"
+	}
+
+	// Check if parent inference service has the status URL
+	if isvc.Status.URL != nil {
+		// Update status to contain the isvc URL with /v1/models/trained-model-name:predict appened
+		url := isvc.Status.URL.String() + "/" + protocolVersion + "/models/" + desiredModel.Name + endpointSuffix
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/trainedmodel/controller.go:156]: We can use [this utility](https://github.com/kubeflow/kfserving/blob/master/pkg/constants/constants.go#L282) to construct the predict path.

**Pull request merged**